File: in_post_BT.java
// Time Complexity :O(n) n-number of nodes
// Space Complexity :O(h) h- maximum height of the tree 
// Did this code successfully run on Leetcode :yes 
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int index;
    HashMap<Integer,Integer> map=new HashMap<>();
    
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        index=postorder.length-1;
        for(int i=0;i<inorder.length;i++)
        {
         map.put(inorder[i],i);   
        }
        
    return helper(inorder,postorder,0,postorder.length-1);
    }
    public TreeNode helper(int[] inorder,int[] postorder,int start,int end)
    {
        if(start>end || index<0) return null;
        TreeNode root=new TreeNode(postorder[index]);
        int inInd=map.get(root.val);
        index--;
       
        root.right=helper(inorder,postorder,inInd+1,end);
        root.left=helper(inorder,postorder,start,inInd-1);
        
        return root;
    }
}

File: InPostBT.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode BuildTree(int[] inorder, int[] postorder) {
        if(inorder.Length == 0) return null;
        var rootVal = postorder[postorder.Length -1];
        var rootIdx = -1;
        var root = new TreeNode(rootVal);
        for(int i = 0; i< inorder.Length; i++)
        { 
            if(inorder[i] == rootVal)
            {
                rootIdx = i;
                break;
            }
        }           
        
        int[] inleft = new int[rootIdx];
        Array.Copy(inorder, 0, inleft, 0, rootIdx);
        int[] inright = new int[inorder.Length - (rootIdx +1)];
        Array.Copy(inorder, rootIdx + 1, inright, 0, inorder.Length - (rootIdx +1));
        int[] postright = new int[inright.Length];
        Array.Copy(postorder, postorder.Length -1 - inright.Length, postright, 0, inright.Length);
        int[] postleft = new int[postorder.Length - postright.Length - 1];
        Array.Copy(postorder, 0, postleft, 0, postorder.Length - postright.Length - 1);
        
        root.left = BuildTree(inleft, postleft);
        root.right = BuildTree(inright, postright);

        return root;        
    }
}
